Add doc examples using attach_hook for code organization instead of LiveComponents (#3685)
phoenixframework/phoenix_live_view: guides/introduction/welcome.md

LiveView supports two extension mechanisms: function components, provided by
`HEEx` templates, and stateful components, known as LiveComponents.

> ### Function Components to organize markup

Reviewer: ```suggestion
### Function components to organize markup
```

module. At the end of the day, they are a useful mechanism for code organization
and to reuse markup in your LiveViews.

Sometimes you need to share more than just markup across LiveViews. When you also

> want to move events to a seperate module, or use the same event handler in multiple

Reviewer 2: ```suggestion
want to move events to a separate module, or use the same event handler in multiple
```

places, function components can be paired with

> [`Phoenix.LiveView.attach_hook/4`](`Phoenix.LiveView.attach_hook/4#sharing-event-handling-logic`)

Reviewer: ```suggestion
[`Phoenix.LiveView.attach_hook/4`](`Phoenix.LiveView.attach_hook/4#sharing-event-handling-logic`).
```

### Live components to encapsulate additional state

A component will occasionally need control over not only its own events,

> but also its own seperate state. For these cases, LiveView

Reviewer: ```suggestion
but also its own separate state. For these cases, LiveView
```

          |> attach_hook(:sort, :handle_event, &MySortComponent.hooked_event/3)  # 2) Delegated events
        {:ok, socket}
      end

      # 1) Normal event
      def handle_event("inc", _params, socket) do
        {:noreply, update(socket, :counter, &(&1 + 1))}
      end
    end

    defmodule MySortComponent do
      use Phoenix.Component

      def display(assigns) do
        ~H"""
        <div :for={{key, list} <- @lists}>
          <ul><li :for={item <- list}>{item}</li></ul>
          <button phx-click="shuffle" phx-value-list={key}>Shuffle</button>
          <button phx-click="sort" phx-value-list={key}>Sort</button>
        </div>
        """
      end

      def hooked_event("shuffle", %{"list" => key}, socket) do
        key = String.to_existing_atom(key)
        shuffled = Enum.shuffle(socket.assigns[key])

        {:halt, assign(socket, key, shuffled)}
      end

      def hooked_event("sort", %{"list" => key}, socket) do
        key = String.to_existing_atom(key)
        sorted = Enum.sort(socket.assigns[key])

        {:halt, assign(socket, key, sorted)}
      end

      def hooked_event(_event, _params, socket), do: {:cont, socket}
    end

> ### Live Components to encapsulate additional state

Reviewer: ```suggestion
### Live components to encapsulate additional state
```

> A component will occasionally need control over not only it's own events,
> but also it's own seperate state. For these cases, LiveView

Reviewer: ```suggestion
A component will occasionally need control over not only its own events,
but also its own seperate state. For these cases, LiveView
```

module. At the end of the day, they are a useful mechanism for code organization
and to reuse markup in your LiveViews.

### attach_hook/4 to organize event handling

However, sometimes you need to share more than just markup across LiveViews,
and you also need to move events to a separate module. For these cases you
can either use regular old functions to delegate, or make use of [attach_hook/4](`Phoenix.LiveView.attach_hook/4#extracting-events-for-code-organization`)

>     defmodule Helpers do

Reviewer 3: I think these helpers detract from the focus of the section. I think it's more explicit to call `{:ok, socket}` and friends, as this is the response people expect from these particular functions.

Author: Thanks. I kind of add those to every LiveView project without thinking, because I hate breaking up the pipeline with that final tuple

That is pretty non-standard though, and you're right, distracts from the main focus here. I'll def swap it to the normal direct tuples

      def halt(socket), do: {:halt, socket}
      def cont(socket), do: {:cont, socket}
    end

    defmodule DemoLive do
      use Phoenix.LiveView
      import Helpers

      def render(assigns) do
        ~H"""
        <div>
          <div>
            Counter: {@counter}
            <button phx-click="inc">+</button>
          </div>

          <MySortComponent.display lists={[first_list: @first_list, second_list: @second_list]} />
        </div>
        """
      end

      def mount(_params, _session, socket) do
        first_list = for(i <- 1..9, do: "First List #{i}") |> Enum.shuffle()
        second_list = for(i <- 1..9, do: "Second List #{i}") |> Enum.shuffle()

        socket
        |> assign(:counter, 0)
        |> assign(first_list: first_list)
        |> assign(second_list: second_list)
        |> attach_hook(:sort, :handle_event, &MySortComponent.hooked_event/3)  # 3) Delegated event using attach_hook
        |> ok
      end

      # 1) Normal event
      def handle_event("inc", _params, socket) do
        socket
        |> update(:counter, &(&1 + 1))
        |> noreply
      end

      # 2) Delegated event using functions

>       def handle_event("shuffle", params, socket), do: MySortComponent.handle_event("shuffle", params, socket)

Reviewer 3: I'm not sure if this is a pattern we want to promote. People are free to do this of course, but it means that changes in the underlying component also require changes to the caller.

Imagine if we changed the event name in our component to "randomize" instead of "shuffle". If we used `attach_hook` and handled the "shuffle" event there, we wouldn't only need to change the `phx-click="shuffle" and `hooked_event("shuffle"...`. and any LiveView depending on this component would continue to function.

If we have delegating using functions, each caller also needs to be updated:

```diff
-      def handle_event("shuffle", params, socket), do: MySortComponent.handle_event("shuffle", params, socket)
+      def handle_event("randomize", params, socket), do: MySortComponent.handle_event("randomize", params, socket)
```

It is also helpful in the case where a component is updated to add new events. Perhaps we want to add a reverse sort button. We can just add it to the component and it'll "just work".

Author: That's a really good point and suddenly makes me look at `attach_hook` in a new light.

I was looking at it as a worse version of defdelegate; but it's really handing off control of that piece of the page. "Here component, you're fully in charge of this now. As you should be"

for creating completely isolated UI elements, but it is a slightly expensive
abstraction if all you want is to compartmentalize markup or events (or both).

### Summary

>   * use `Phoenix.Component` for code organization and reusing markup
>   * pair `Phoenix.Component` with [`attach_hook/4`](`Phoenix.LiveView.attach_hook/4#sharing-event-handling-logic`) to organize and reuse event handling

Reviewer 2: ```suggestion
  * use `Phoenix.Component` for code organization and reusing markup (optionally with [`attach_hook/4`](`Phoenix.LiveView.attach_hook/4#sharing-event-handling-logic`) for event handling reuse)
```

for creating completely isolated UI elements, but it is a slightly expensive
abstraction if all you want is to compartmentalize markup or events (or both).

### Summary

In LiveViews we may want to organize code by extracting related sections to another module

There are four common situations
  1. Extract html

>   2. Extract html and related handle_events
>   3. Extract html, handle_events, and create additional separate state

Reviewer: ```suggestion
  2. Extract html and related events
  3. Extract html, events, and create additional separate state
```

  4. Same as 3, but with error isolation

>   For 1, we have [FunctionComponents](`Phoenix.Component`)

Reviewer: ```suggestion
  For 1, we have [function components](`Phoenix.Component`)
```

>   For 2, we can combine FunctionComponents & [attach_hook/4](`Phoenix.LiveView.attach_hook/4`)

Reviewer: ```suggestion
  For 2, we can combine function components & [attach_hook/4](`Phoenix.LiveView.attach_hook/4`)
```

>   Situation 3 is covered by [LiveComponents](`Phoenix.LiveComponent`)

Reviewer: ```suggestion
  Situation 3 is covered by [live components](`Phoenix.LiveComponent`)
```
